editor/src/components/navigator/navigator-utils.ts

```diff
@@ -79,317 +106,677 @@ export function navigatorDepth(
   return result
 }
 
+type RegularNavigatorTree = {
+  type: 'regular-entry'
+  elementHidden: boolean
+  subtreeHidden: boolean
+  navigatorEntry: NavigatorEntry
+  renderProps: { [propName: string]: NavigatorTree }
+  children: Array<NavigatorTree>
+}
+
+// maybe the leaf is not actually useful and we can remove it
+type LeafNavigatorTree = {
+  type: 'leaf-entry'
+  elementHidden: boolean
+  navigatorEntry: NavigatorEntry
+}
+
+type MapNavigatorTree = {
+  type: 'map-entry'
+  elementHidden: boolean
+  subtreeHidden: boolean
+  navigatorEntry: NavigatorEntry
+  mappedEntries: Array<NavigatorTree>
+}
+
+type ConditionalNavigatorTree = {
+  type: 'conditional-entry'
+  elementHidden: boolean
+  subtreeHidden: boolean
+  navigatorEntry: NavigatorEntry
+  trueCase: Array<NavigatorTree>
```

Comment:
Do we really need an array here? Conditional branches contain a single element.

Reply:
the Conditional branch being a single element is a limitation of the parser, felt easier and cheaper to not assume that it can only be a single element (and then put in guards to ensure that we never encounter a conditional with multiple elements)